### PR TITLE
FMS2_io: Boundary condition restarts and the get/set_filename appendix

### LINKED ISSUE #543: FMS2: Need an equivalent to register_restart_region_r2d (and 3d)
**Is your feature request related to a problem? Please describe.**
The GFDL_atmos_cubed_sphere file [tools/fv_io.F90](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/master/tools/fv_io.F90#L678) uses the ```fms_io_mod``` function ```register_restart_field```.  I am in the process of updating this to use the ```fms2_io_mod``` function ```register_restart_field``` .  ```register_restart_field``` in ```fms_io_mod``` is an interface pointing to multiple subroutine procedures, 2 of which are ```register_restart_region_r2d``` and ```register_restart_region_r3d```.   These are the subroutines being called in [tools/fv_io.F90](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/blob/master/tools/fv_io.F90#L678).

**Describe the solution you'd like**
I would like subroutine to be added to the ```fms2_io_mod``` ```register_restart_field``` interface that do what [register_restart_region_r2d](https://github.com/NOAA-GFDL/FMS/blob/master/fms/fms_io.F90#L2373) and [register_restart_region_r3d](https://github.com/NOAA-GFDL/FMS/blob/master/fms/fms_io.F90#L2428) do.

**Describe alternatives you've considered**
None.

**Additional context**
None.

### LINKED ISSUE #545: FMS2: Need equivalent for set_filename_appendix and get_filename_appendix in fms2_io_mod
**Is your feature request related to a problem? Please describe.**
As discussed in the FMS I/O Meeting on July 20th, the following subroutines are in ```fms_io_mod```. As we move from the use of ```fms_io_mod``` to ```fms2_io_mod``` in GFDL_atmos_cubed_sphere and MOM6, these functions will need to be used still and are not in fms2_io_mod.
[set_filename_appendix](https://github.com/NOAA-GFDL/FMS/blob/master/fms/fms_io.F90#L8382)
[get_filename_appendix](https://github.com/NOAA-GFDL/FMS/blob/master/fms/fms_io.F90#L8366)

**Describe the solution you'd like**
The group decided that these functions should be replicated in ```fms2_io_mod```.

**Describe alternatives you've considered**
None.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**

Adds support for boundary condition restarts and the get/set_filename appendix
Adds unit tests

Fixes #545 
Fixes #543

**How Has This Been Tested?**
`make check` passes with intel/gcc

Needs a model test. 

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

**Review by laurenchilutti — DISMISSED** [review]

**Comment by thomas-robinson** [comment]
@uramirez @rem1776 is this getting pushed to 2021.02?

**Comment by uramirez8707** [comment]
@thomas-robinson yes 2021.02

**Review by laurenchilutti — DISMISSED** [review]

File: fms2_io/netcdf_io.F90 [hunk 1]
```diff
@@ -1756,7 +1815,8 @@ end subroutine compressed_start_and_count
 include "get_variable_attribute.inc"
 include "compressed_write.inc"
 include "compressed_read.inc"
-
+#include "scatter_data_bc.inc"
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply fms2_io/netcdf_io.F90]: Yes, the # is not needed. Thanks!

**Review by laurenchilutti — APPROVED** [review]

**Review by bensonr — DISMISSED** [review]

**Review by wrongkindofdoctor — DISMISSED** [review]

**Review by rem1776 — DISMISSED** [review]

File: fms2_io/include/gather_data_bc.inc [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+       deallocate(local_buf_r8_kind)
+       !> If you are on fileobj's root, calculate the checksum and save the gathered data in a buffer
+       if (fileobj%is_root) then
+           chksum_val = mpp_chksum(global_buf_r8_kind(1:i_glob,1:j_glob), (/mpp_pe()/))
+           allocate(bc_info%globaldata2d_r8(i_glob, j_glob))
+           bc_info%globaldata2d_r8=global_buf_r8_kind(1:i_glob,1:j_glob)
+           deallocate(global_buf_r8_kind)
+       endif
+     class default
+        call error("unsupported type.")
```
Comment (thomas-robinson) [inline fms2_io/include/gather_data_bc.inc:125]: @uramirez8707 This error message needs to be updated

File: fms2_io/include/gather_data_bc.inc [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+       deallocate(local_buf_r8_kind)
+       !> If you are on fileobj's root, calculate the checksum and save the gathered data in a buffer
+       if (fileobj%is_root) then
+           chksum_val = mpp_chksum(global_buf_r8_kind(1:i_glob,1:j_glob, :), (/mpp_pe()/))
+           allocate(bc_info%globaldata3d_r8(i_glob, j_glob, bc_info%global_size(3)))
+           bc_info%globaldata3d_r8=global_buf_r8_kind(1:i_glob,1:j_glob,:)
+           deallocate(global_buf_r8_kind)
+       endif
+     class default
+        call error("unsupported type.")
```
Comment (thomas-robinson) [inline fms2_io/include/gather_data_bc.inc:246]: This error message needs to more descriptive

File: fms2_io/include/gather_data_bc.inc [hunk 4]
```diff
@@ -0,0 +1,256 @@
+!* This file is part of the GFDL Flexible Modeling System (FMS).
```
Comment (thomas-robinson) [inline fms2_io/include/gather_data_bc.inc:1]: The top of the license header is missing?

File: fms2_io/netcdf_io.F90 [hunk 5]
```diff
@@ -61,9 +81,10 @@ module netcdf_io_mod
   class(*), dimension(:,:,:,:,:), pointer :: data5d => null() !< 5d data pointer.
   logical :: was_read !< Flag to support legacy "query_initialized" feature, which
                       !! keeps track if a file was read.
+  logical :: is_bc_variable
+  type(bc_information) :: bc_info
```
Comment (thomas-robinson) [inline fms2_io/netcdf_io.F90:85]: These variables should have descriptions

File: fms2_io/netcdf_io.F90 [hunk 6]
```diff
@@ -456,7 +502,7 @@ function netcdf_file_open(fileobj, path, mode, nc_format, pelist, is_restart, do
 
   integer :: nc_format_param
   integer :: err
-  character(len=256) :: buf
+  character(len=256) :: buf, buf2
```
Comment (thomas-robinson) [inline fms2_io/netcdf_io.F90:505]: Why is there a buf and a buf2.  Please document.

File: test_fms/fms2_io/test_bc_restart.F90 [hunk 7]
```diff
@@ -0,0 +1,277 @@
+!* This file is part of the GFDL Flexible Modeling System (FMS).
```
Comment (thomas-robinson) [inline test_fms/fms2_io/test_bc_restart.F90:1]: The top of the license header is missing

**Review by thomas-robinson — CHANGES_REQUESTED** [review]
Please fix headers and document the `buf2` variable.

**Review by thomas-robinson — CHANGES_REQUESTED** [review]
Please fix headers and document the `buf2` variable.

File: fms2_io/include/gather_data_bc.inc (hunk 2 above)
Comment (uramirez8707) [reply fms2_io/include/gather_data_bc.inc]: Fixed in f0d7161f31478f7ef7f92b4eee505d409ef3b0eb

File: fms2_io/include/gather_data_bc.inc (hunk 3 above)
Comment (uramirez8707) [reply fms2_io/include/gather_data_bc.inc]: Fixed in f0d7161f31478f7ef7f92b4eee505d409ef3b0eb

File: fms2_io/include/gather_data_bc.inc (hunk 4 above)
Comment (uramirez8707) [reply fms2_io/include/gather_data_bc.inc:4]: Fixed in f0d7161f31478f7ef7f92b4eee505d409ef3b0eb

File: fms2_io/netcdf_io.F90 (hunk 5 above)
Comment (uramirez8707) [reply fms2_io/netcdf_io.F90]: Fixed in f0d7161f31478f7ef7f92b4eee505d409ef3b0eb

File: fms2_io/netcdf_io.F90 (hunk 6 above)
Comment (uramirez8707) [reply fms2_io/netcdf_io.F90]: Fixed in f0d7161f31478f7ef7f92b4eee505d409ef3b0eb

File: test_fms/fms2_io/test_bc_restart.F90 (hunk 7 above)
Comment (uramirez8707) [reply test_fms/fms2_io/test_bc_restart.F90:4]: Fixed in f0d7161f31478f7ef7f92b4eee505d409ef3b0eb

**Review by thomas-robinson — APPROVED** [review]

**Review by laurenchilutti — APPROVED** [review]

**Review by wrongkindofdoctor — DISMISSED** [review]

**Review by bensonr — COMMENTED** [review]
I'm a little confused at the approach taken here for reading and writing BCs.  For ingesting the BC data, you have read_restart_bc simply call scatter_data_bc_*, which performs both the read and data scatter from within that routine.  Yet for outputting the BC data, the write_restart_bc calls gather_data_bc_* to merely gather the data to the io_root and then return control back to write_restart_bc write the data.  It would make sense to have read/write_restart_bc and the resulting gather/scatter_data_bc perform in the same manner.

**Comment by uramirez8707** [comment]
>For ingesting the BC data, you have read_restart_bc simply call scatter_data_bc__, which performs both the read and data scatter from within that routine. 

This cannot really happen with the writes. `gather_data_bc` will have to gather the data, calculate the checksum, write the checksum and the data. Because it is in a do loop per variable, you will be switching between writing data and meta data after every variable and if the header is not big enough netcdf will probably have to rewrite the file. 

>Yet for outputting the BC data, the write_restart_bc calls gather_data_bc__ to merely gather the data to the io_root and then return control back to write_restart_bc write the data. It would make sense to have read/write_restart_bc and the resulting gather/scatter_data_bc perform in the same manner.

I think this will be messy for the reads. `read_restart_bc` will have to allocate space for the global_data (which will require a `select type` for each dimension), read into that buffer and then call scatter_data_bc to scatter it.

**Comment by bensonr** [comment]
@uramirez8707 - thanks for the explanation (makes perfect sense).  Can you put some inline comments in the code as to why things are structured in this way?

**Review by rem1776 — APPROVED** [review]

**Review by laurenchilutti — APPROVED** [review]

**Review by thomas-robinson — APPROVED** [review]
